wine-app-packager.1
.TH WINE-APP-PACKAGER "1" "June 2009" "wine-app-packager 0.1" "User Commands"
.SH NAME
wine-app-packager \- helps you creating Debian packages for Windows programs
.SH SYNOPSIS
.B wine-app-packager\fR init
.PP
.B wine-app-packager\fR prepare
.PP
.B wine-app-packager\fR run \fI<cmd>\fR
.PP
.B wine-app-packager\fR commit
.PP
.B wine-app-packager\fR abort
.PP
.SH DESCRIPTION
.B wine-app-packager
guides you in wrapping a Windows program in a Debian package,
based on wine. It will create a template Debian source package for you and
include the \fIC:\\\fR directory tree of your program.
.PP
The produced packages will install the provided \fIdrive_c/\fR directory  and a
copy of the wine configuration to the appropriate places (see below). A wrapper
will be placed in \fI/usr/bin/<application>\fR. This wrapper also contains the
name of the actual \fI.exe\fR file to be used.
.PP
The package will depend on the
currently used wine package, i.e. the package that ships the wine binary configured in the the wrapper script (default: \fI/usr/bin/wine\fR). If that file happens to be a symlink, it will be followed.

.SH OPTIONS
.TP
.B wine-app-packager\fR init
creates a Debian source package directory. It will ask various pieces
of information. All the following commands need to be run within the
directory \fI<application>-<version>/\fR that is now created in the current
directory.

.TP
.B wine-app-packager\fR prepare
creates a wine config directory in \fI~/.wine-<application>-master\fR, based on
the current content of the package (if any).

.TP
.B wine-app-packager\fR run \fI<cmd>\fR
runs \fR<cmd>\fR in a way that \fB$WINEPREFIX\fR is set correctly, i.e.
.nf
wine-app-packager run wine /tmp/install.exe
.fi
or
.nf
wine-app-packager run wine C:\\\\Applications\\\\DemoApplication\\\\DemoApplication.exe
.fi
If you want to use a wine installation that is not in the default path, you
need to specify the full path to the wine binary here. In this case, you also
need to update the \fB$WINE\fR variable in the wrapper script.

.TP
.B wine-app-packager\fR commit
takes the changes you made in \fI~/.wine-<application>-master\fR and puts them
back into the Debian source package. Usually, this will be followed by
calls to \fBdebchange\fR and \fBdebuild\fR to build the package.
Removes \fI~/.wine-<application>-master\fR afterwards.

.TP
.B wine-app-packager\fR abort
just removes \fI~/.wine-<application>-master\fR, throwing away all changes.

.SH "FILES"

The resulting \fBwine-<application>\fR packages use the following files and directories:

.TP
\fI/usr/bin/<application>\fR
Shell wrapper that creates the users dedicated \fB$WINEPREFIX\fR as
\fI~/.wine-<application>\fR, fills it with the provided wine configuration and
symlinks to the \fIdrive_c/\fR and starts the application, using the configured wine
binary (default: \fI/usr/bin/wine\fR).

.TP
\fI/opt/wineapps/wine-<application>\fR
The location for the prepared \fIC:\\\fR folder.

.TP
\fI/opt/wineapps/wine-<application>/wine-config.tar.gz\fR
The template for the users wine configuration directory.

.SH "EXAMPLES"

A typical run of \fBwine-app-packager\fR that creates a new package, builds it and run the program would look like this (only user input shown):
.PP
.nf
$ wine-app-packager init
7z
1.2.3
$ cd 7z-1.2.3
$ wine-app-packager prepare
$ wine-app-packager run /opt/wine/wine-3.4.5/bin/wine /tmp/7z465.exe
$ wine-app-packager commit
$ edit ./z7 # set the EXE= variable, possibly the WINE= variable, too
$ debuild -uc -us
$ sudo debi
$ 7z
.fi

.SH AUTHOR
Joachim Breitner <joachim.breitner@itomig.de>



debian/changelog
wine-app-packager (0.1) UNRELEASED; urgency=low

  * Initial release

 -- Joachim Breitner <joachim.breitner@itomig.de>  Fri, 12 Jun 2009 11:12:54 +0200


wine-app-packager.sh
#!/bin/bash

# TODO:
# * LHM-Wine-Pakete beachten

set -e

#
# © 2009 Joachim Breitner <joachim.breitner@itomig.de>
#
# Licensed under the EUPL, Version 1.0 or – as soon they
# will be approved by the European Commission - subsequent
# versions of the EUPL (the "Licence");
# you may not use this work except in compliance with the
# Licence.
# You may obtain a copy of the Licence at:
# 
# http://ec.europa.eu/idabc/7330l5
# 
# Unless required by applicable law or agreed to in
# writing, software distributed under the Licence is
# distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
# express or implied.
# See the Licence for the specific language governing
# permissions and limitations under the Licence.

function common_setup () {
	if ! [ -d debian ]
	then
		print "Please run command $0 $1 from within the source directory"
		exit 1;
	fi

	APPNAME="$(grep-dctrl -n -F Source -s Source '' < debian/control)"
	DEBVERSION="$(dpkg-parsechangelog | grep-dctrl -n -s Version '')"
	WINE_MASTER_DIR=~/".wine-$APPNAME-master"
}

if [ ! "$1" ]
then
	cat <<'__END__'
wine-app-packager
=================

This script aids you in creation of Debian packages of Windows applications
that are meant to run under wine. It offers the following commands:

wine-app-packager init
	creates a Debian source package directory. It will ask various pieces
	of information. All the following commands will need to run within the
	created directory.

wine-app-packager prepare
	Creates a wine config directory in ~/.wine-<application>-master, based on
	the current content of the package (if any).

wine-app-packager run <cmd>
	Runs cmd in a way that WINEPREFIX is set correctly, i.e.
	wine-app-packager run wine /tmp/install.exe
	or
	wine-app-packager run wine C:\\Applications\\DemoApplication\\DemoApplication.exe

wine-app-packager commit
	Takes the changes you made in ~/.wine-<application>-master and puts them
	back into the Debian source package. Ususally, this will be followed by
	calls to debchange and debuilt to build the package
	Removes ~/.wine-<application>-master afterwards.

wine-app-packager abort
	Just removes ~/.wine-<application>-master, throwing away all changes.
__END__
	exit 1;
fi

if [ "$1" = "init" ]
then
	cat <<__END__
Please enter the name of the application (only lower letters, digits, dots and
dashes. It will be both the package name and the name of the wrapper executable.
Example: wsftp
__END__

	read APPNAME
	if [ -z "$APPNAME" -o -n "$(echo -n "$APPNAME" | tr -d a-z0-9.-)" ]
	then
		echo "Invalid application name \"$APPNAME\"."
		exit 1
	fi

	cat <<__END__
Please enter the upstream version of the application, with the same conventions.
If there is no meaningful version, leave this field empty.
Example: 1.0
__END__

	read APPVER
	if [ -n "$(echo -n "$APPVER" | tr -d a-z0-9.-)" ]
	then
		echo "Invalid application version \"$APPVER\"."
		exit 1
	fi

	if [ -z "$APPVER" ]
	then
		DEBVER="1"
	else
		DEBVER="$APPVER"
	fi
	DEBDIR="$APPNAME-$DEBVER"

	if [ -e "$DEBDIR" ]
	then
		cat <<__END__
The directory \"$DEBDIR\" does already exist. Please remove it if you want to
start from scratch.
__END__

		exit 1
	fi
	echo "Creating directory \"$DEBDIR\"."
	mkdir "$DEBDIR"
	mkdir "$DEBDIR"/debian
	echo "Writing \"$DEBDIR/debian/control\""

	if [ -n "$DEBFULLNAME" ]
	then
		MAINT="$DEBFULLNAME <$DEBEMAIL>"
	else
		cat <<__END__
WARNING: DEBFULLNAME and DEBEMAIL not set. Please adjust maintainer field
in debian/control and debian/changelog afterwards.
__END__
		MAINT="Someone <please-update@in.here>"
	fi
	cat >"$DEBDIR/debian/control" <<__END__
Source: $APPNAME
Maintainer: $MAINT

Package: $APPNAME
Architecture: all
Depends: wine
Description: Application $APPNAME
 This is a windows application packaged for Debian clients.
__END__
	echo "Writing \"$DEBDIR/debian/compat\""
	echo 5 > "$DEBDIR/debian/compat"
	echo "Writing \"$DEBDIR/debian/rules\""
	cat >"$DEBDIR/debian/rules" <<__END__
#!/usr/bin/make -f
build:
	

clean:
	dh_testdir
	dh_testroot
	dh_clean

install: build
	dh_testdir
	dh_testroot
	dh_clean -k
	dh_installdirs

	# Add here commands to install the package into debian/<packagename>
	#\$(MAKE) prefix=\`pwd\`/debian/\`dh_listpackages\`/usr install

binary-indep: install

# Build architecture-dependent files here.
binary-arch: install
	dh_testdir
	dh_testroot
	dh_installchangelogs
	dh_installdocs
	dh_installexamples
	dh_install
	dh_installman
	dh_link
	dh_strip
	dh_compress
	dh_fixperms
	dh_installdeb
	dh_shlibdeps
	dh_gencontrol
	dh_md5sums
	dh_builddeb

binary: binary-indep binary-arch
.PHONY: clean binary-indep binary-arch binary install
__END__
	chmod +x "$DEBDIR/debian/rules"
	echo "Generating \"$DEBDIR/debian/install\""
	cat >"$DEBDIR/debian/install" <<__END__
drive_c /opt/wineapps/wine-$APPNAME/
wine-config.tar.gz /opt/wineapps/wine-$APPNAME/
$APPNAME /usr/bin
__END__
	echo "Generating \"$APPNAME\" (start wrapper. Please adjust variables at the top)"
	cat >"$DEBDIR/$APPNAME" <<__END__
#!/bin/bash

# This variable need to be set:
EXE='c:\\Programme\\Please Adjust\\start.exe'

# These are already filled correctly
APPNAME="$APPNAME"
WINE_HOME="\$HOME/.wine-\$APPNAME"

set -e
if [ ! -d "\$WINE_HOME" ]
then
	echo "\$APPNAME started for the first time, creating \$WINE_HOME"
	mkdir -p "\$WINE_HOME"
	tar --extract --gzip --file /opt/wineapps/wine-\$APPNAME/wine-config.tar.gz -C "\$WINE_HOME"
	cp -r --symbolic-link /opt/wineapps/wine-\$APPNAME/drive_c/ "\$WINE_HOME"/drive_c
fi


export WINEPREFIX="\$WINE_HOME"
exec wine "\$EXE"
__END__
	chmod +x "$DEBDIR/$APPNAME" 
	echo "Generating \"$DEBDIR/debian/changelog\""
	pushd "$DEBDIR"
 	debchange --create --package "$APPNAME" --newversion "$DEBVER-1" --distribution UNRELEASED "First release of $APPNAME"
	popd
	cat <<__END__
Done preparing the source package. You can review the contents of these files
now, or go ahead with wine-app-packager prepare, from within the created source
directory in $DEBDIR.
__END__

elif [ "$1" = "prepare" ]
then
	common_setup

	if [ -e "$WINE_MASTER_DIR" ]
	then
		cat <<__END__
Directory $WINE_MASTER_DIR already exists. If it contains changes
that need to be preserved, please use $0 commit, otherwise remove
it before running $0 $1.
__END__
		exit 1
	fi

	if [ -d drive_c ]
	then
		echo "Creating $WINE_MASTER_DIR"
		mkdir "$WINE_MASTER_DIR"
		echo "Extracting ./wine-config.tar.gz to $WINE_MASTER_DIR"
		tar --extract --gzip --file ./wine-config.tar.gz -C "$WINE_MASTER_DIR"
		echo "Copying ./drive_c to $WINE_MASTER_DIR/drive_c"
		rsync --recursive ./drive_c/ $WINE_MASTER_DIR/drive_c
	else
		echo "Empty package, creating empty $WINE_MASTER_DIR"
		mkdir "$WINE_MASTER_DIR"
	fi
elif [ "$1" = "run" ]
then
	common_setup

	if [ ! -d "$WINE_MASTER_DIR" ]
	then
		cat <<__END__
Directory $WINE_MASTER_DIR does not exist. Please run $0 $1 prepare first.
__END__
		exit 1
	fi

	if [ -z "$2" ]
	then
		cat <<__END__
You need to pass a program to be run, e.g. $0 $1 wine setup.exe
__END__
		exit 1
	fi

	export WINEPREFIX=$WINE_MASTER_DIR
	shift
	exec "$@"
elif [ "$1" = "commit" ]
then
	common_setup

	if [ ! -d "$WINE_MASTER_DIR" ]
	then
		cat <<__END__
Directory $WINE_MASTER_DIR does not exist. Please run $0 $1 prepare first.
__END__
		exit 1
	fi

	echo "Copying over $WINE_MASTER_DIR in a tarfile, excluding drive_c"
	tar --create --file ./wine-config.tar.gz --gzip --directory "$WINE_MASTER_DIR" --exclude drive_c .
	echo "Copying over drive_c, excluding Profiles"
	rsync --recursive --delete --exclude windows/profiles --exclude windows/temp "$WINE_MASTER_DIR/drive_c/" drive_c

	echo "Removing $WINE_MASTER_DIR"
	rm -rf "$WINE_MASTER_DIR"
	cat <<__END__
Change are propagated to the Debian source directory. To build the package,
increase the version number with "debchange -i" and build it with "dpkg-buildpackage -uc -us".
__END__

elif [ "$1" = "abort" ]
then
	common_setup

	if [ ! -d "$WINE_MASTER_DIR" ]
	then
		cat <<__END__
Directory $WINE_MASTER_DIR does not exist, nothing to abort.
__END__
		exit 1
	fi

	echo "Removing $WINE_MASTER_DIR"
	rm -rf "$WINE_MASTER_DIR"
else
	echo "Unknown command $1. Please run $0 without commands for a usage summary."
	exit 1
fi
